File: container2/leetcodely/python/number_of_islands.py
"""Given a 2d grid map of '1's (land) and '0's (water), count the number of islands. An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.
"""


class Solution(object):

    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        def dfs(grid, r, c):
            grid[r][c] = '0'
            for (dx, dy) in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                nx, ny = r + dx, c + dy
                if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and grid[nx][ny] == '1':
                    dfs(grid, nx, ny)
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '1':
                    dfs(grid, i, j)
                    count += 1
        return count


if __name__ == '__main__':
    grid = [['1', '1', '0', '0', '0'],
            ['1', '1', '0', '0', '0'],
            ['0', '0', '1', '0', '0'],
            ['0', '0', '0', '1', '1']
            ]
    s = Solution()
    print(s.numIslands(grid))

File: container2/leetcodely/java/src/io/soumasish/NumberOfIslands.java
package io.soumasish;

public class NumberOfIslands {
    private int[] dx = {0, 0, -1, 1};
    private int[] dy= {1, -1, 0, 0};

    public int numIslands(char[][] grid) {
        int count = 0;
        if(grid.length == 0 || grid[0].length == 0) return 0;
        for (int i = 0; i <grid.length ; i++) {
            for (int j = 0; j <grid[i].length ; j++) {
                if (grid[i][j] == '1'){
                    count += 1;
                    dfs(grid, i, j);
                }
            }
        }
        return count;

    }
    public void dfs(char[][] grid, int i, int j){
        if(i < 0 || i > grid.length || j < 0 || j > grid[0].length || grid[i][j] == '0') return;
        grid[i][j] = '0';
        for (int k = 0; k <4 ; k++) {
            int nx = i + dx[k];
            int ny = j + dy[k];
            dfs(grid, nx, ny);

        }

    }
}
